### PR TITLE
Enable warnings as errors in compiler component on x86

### PR DESCRIPTION
**Pull request by fjeremic**
A series of commits which address warnings on Windows and Linux on x86, following which we enable warnings as errors on all x86 platforms.

### CODE REVIEW

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/optimizer/VPHandlers.cpp
Comment (fjeremic) [inline compiler/optimizer/VPHandlers.cpp:4761]: This to me looks like an actual bug, so pointing it out here. Here is the warning:

```
[2021-05-28T22:32:43.924Z]     32>C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4604): warning C4244: 'argument' : conversion from 'int64_t' to 'int32_t', possible loss of data [C:\jenkins\workspace\PullRequest-win_x86-64\Build\build\jitbuilder\jitbuilder.vcxproj]
[2021-05-28T22:32:43.924Z]                  C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4678) : see reference to function template instantiation 'void addValidRangeBlockOrGlobalConstraint<TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),int32_t(__cdecl *)(int64_t),T>(OMR::ValuePropagation *,TR::Node *,FUNC,FUNC2,T,T,bool)' being compiled
[2021-05-28T22:32:43.924Z]                  with
[2021-05-28T22:32:43.924Z]                  [
[2021-05-28T22:32:43.924Z]                      T=int64_t,
[2021-05-28T22:32:43.924Z]                      FUNC=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),
[2021-05-28T22:32:43.924Z]                      FUNC2=int32_t (__cdecl *)(int64_t)
[2021-05-28T22:32:43.924Z]                  ]
[2021-05-28T22:32:43.924Z]                  C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4761) : see reference to function template instantiation 'TR::Node *constrainHighestOneBitAndLeadingZerosHelper<TR::VPLongConst*(__cdecl *)(TR::VPConstraint *),TR::VPLongRange*(__cdecl *)(TR::VPConstraint *),void(__cdecl *)(TR::VPLongConst *,int64_t &),void(__cdecl *)(TR::VPLongRange *,int64_t &,int64_t &),TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t),TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),int32_t(__cdecl *)(int64_t),int64_t>(OMR::ValuePropagation *,TR::Node *,A,B,C,D,E,F,G,T,T)' being compiled
[2021-05-28T22:32:43.924Z]                  with
[2021-05-28T22:32:43.924Z]                  [
[2021-05-28T22:32:43.924Z]                      A=TR::VPLongConst *(__cdecl *)(TR::VPConstraint *),
[2021-05-28T22:32:43.924Z]                      B=TR::VPLongRange *(__cdecl *)(TR::VPConstraint *),
[2021-05-28T22:32:43.924Z]                      C=void (__cdecl *)(TR::VPLongConst *,int64_t &),
[2021-05-28T22:32:43.924Z]                      D=void (__cdecl *)(TR::VPLongRange *,int64_t &,int64_t &),
[2021-05-28T22:32:43.924Z]                      E=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t),
[2021-05-28T22:32:43.924Z]                      F=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),
[2021-05-28T22:32:43.924Z]                      G=int32_t (__cdecl *)(int64_t),
[2021-05-28T22:32:43.924Z]                      T=int64_t
[2021-05-28T22:32:43.924Z]                  ]
```

It seems like we're trying to create a constraint with integer range, but the types here are all longs.

File: compiler/infra/InterferenceGraph.cpp
Comment (fjeremic) [inline compiler/infra/InterferenceGraph.cpp:413]: Highlighting this bug as well. I'm failure sure this code is dead though as we haven't done graph colouring in a while.

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/optimizer/DeadTreesElimination.cpp
Comment (fjeremic) [inline compiler/optimizer/DeadTreesElimination.cpp:632]: This is also a bug. We're passing in the block pointer as a boolean which determines whether we want to request this optimization, however we're requesting it on a `NULL` block instead of the block we want it in!

File: compiler/optimizer/LoopVersioner.cpp
Comment (fjeremic) [inline compiler/optimizer/LoopVersioner.cpp:5915]: In this bug we are passing the compilation object pointer as a boolean which determines whether to duplicate children.

File: compiler/x/codegen/SIMDTreeEvaluator.cpp
Comment (fjeremic) [inline compiler/x/codegen/SIMDTreeEvaluator.cpp:45]: This bug forces us to generate 64-bit `lea` instruction always.

File: compiler/optimizer/VPHandlers.cpp
Comment (fjeremic) [inline compiler/optimizer/VPHandlers.cpp:6934]: This seems to have been introduced in #2711. @jdmpapin FYI. I think the change here is what we want. Could you please review this one? Here was the warning:

```
warning C4146: unary minus operator applied to unsigned type, result still unsigned
```

File: compiler/x/amd64/codegen/AMD64SystemLinkage.cpp
Comment (fjeremic) [inline compiler/x/amd64/codegen/AMD64SystemLinkage.cpp:907]: This also seems very dodgy. The `-stackIndex` and `-pos` parts. It seems we are negating unsigned values. Tagging @0xdaryl to review this one. Here was the warning:

```
warning C4146: unary minus operator applied to unsigned type, result still unsigned
```

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/infra/InterferenceGraph.cpp
Comment (Leonardo2718) [reply compiler/infra/InterferenceGraph.cpp:413]: Actually, I believe graph colouring is still used for compacting locals, though it's unclear how much of the code actually gets used:

https://github.com/eclipse/omr/blob/e358cd0e50776aacc01d109deb37c454901dad05/compiler/codegen/CodeGenPrep.cpp#L352-L416

File: compiler/x/amd64/codegen/AMD64SystemLinkage.cpp
Comment (Leonardo2718) [reply compiler/x/amd64/codegen/AMD64SystemLinkage.cpp:907]: This code looks very odd. The `~ + 1` trick might silence the warning, but it doesn't clarify what the code is trying to do. My vote is to rewrite this code to be more sensible (or add a big comment block if this is actually the intended behaviour).

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/x/amd64/codegen/AMD64SystemLinkage.cpp
Comment (fjeremic) [reply compiler/x/amd64/codegen/AMD64SystemLinkage.cpp:907]: I fixed the warning but I'm not sure if the code is correct, so wanted to point it out. I did not change the semantics of the code.

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/optimizer/VPHandlers.cpp
Comment (jdmpapin) [reply compiler/optimizer/VPHandlers.cpp:6934]: The cast I introduced there was definitely intentional:
> Additionally, these handlers will now cast to unsigned before negating when computing the constraint for a constant, in order to ensure that there is no signed overflow.

Removing the cast will cause the (signed) negation to overflow if `value` is `INT32_MIN`, and that overflow is undefined behaviour (though it will probably happen to work fine anyway). And it's similar in `constrainLabs()` just below when `value` is `INT64_MIN`.

I think this is a bad warning in these two cases. It doesn't matter at all that the result is still unsigned, because the only use of the result is to immediately reinterpret it as a signed integer of the same size. And the change that the warning seems to encourage (i.e. removing the cast) reintroduces UB that the cast was introduced expressly to avoid.

It's possible though that appeasing the compiler will still result in cleaner code than suppressing the warning here with an explanation why. In that case, we can remove the cast, but we need to exclude the `INT32_MIN`/`INT64_MIN` cases, e.g.
```
if (value < 0 && value != INT32_MIN)
   value = -value;
```
Because negation doesn't do anything to the minimum value, it's fine to skip it.

**Comment by fjeremic** [comment]
Jenkins build all

File: compiler/optimizer/VPHandlers.cpp
Comment (fjeremic) [reply compiler/optimizer/VPHandlers.cpp]: Thanks for the explanation. I've implemented the fix in a new commit so it is isolated. Fixed in 50d0dbd.

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build all

File: cmake/modules/OmrTargetSupport.cmake
Comment (dnakamura) [inline cmake/modules/OmrTargetSupport.cmake:81]: My concern here is that we lose warnings as errors for all non c/c++ languages, not just `ASM_NASM` eg `ASM` and `ASM_MASM` I think a better solution would be to make `OMR_<LANG>_WARNINGS_AS_ERROR_FLAG` variables and iterate the `ENABLED_LANGUAGES` property to apply all the flags

**Comment by fjeremic** [comment]
Jenkins build all

**Comment by fjeremic** [comment]
Jenkins build osx

**Comment by fjeremic** [comment]
Removing WIP as all issues have been fixed in OMR and downstream projects. Tagging a few people who may be interested in reviewing @dsouzai @0xdaryl @ymanton @jdmpapin. Also I'm looking for a committer for this PR. Thanks all! I know it's a big PR but the changes are separated into logical commits fixing one warning at a time so the changes in each commits are quite similar.

File: cmake/modules/OmrTargetSupport.cmake
Comment (fjeremic) [reply cmake/modules/OmrTargetSupport.cmake:81]: @dnakamura would you be ok if we left this as is and if I opened an issue to track adding `OMR_<LANG>_WARNINGS_AS_ERROR_FLAG` which I will work on immediately after this PR is merged? Reason I'm suggesting this is due to the size of the PR and we just got the builds fully green so I want to merge what we have to enable the warnings as errors to prevent additional warnings being introduced. If you can give me a thumbs up I will open up an issue, assign to myself, and start working on it in the background as this PR gets reviewed.

File: compiler/x/codegen/SIMDTreeEvaluator.cpp
Comment (dsouzai) [reply compiler/x/codegen/SIMDTreeEvaluator.cpp:45]: I think this bug may have had the side effect of doing the right thing for AOT. Because this code is guarded by an "is unresolved" check, you may want to do
```
generateRegMemInstruction(LEARegMem(cg->comp()->compileRelocatableCode()), node, tempReg, mr, cg);
```

Comment (fjeremic) [reply compiler/x/codegen/SIMDTreeEvaluator.cpp:45]: Why would AOT or not matter here? The line being modified here is determining whether we should generate a 64-bit or 32-bit load effective address instruction.

Comment (dsouzai) [reply compiler/x/codegen/SIMDTreeEvaluator.cpp:45]: In AOT, it's always a 64-bit (or I suppose a pointer-sized) load address, because you don't know if the address you want is always going to be 32-bit or 64 bit.

File: compiler/env/FEBase.cpp
Comment (fjeremic) [reply compiler/env/FEBase.cpp]: Not sure what happen here. I've removed the commit in a force-push.

File: compiler/optimizer/OptimizationData.hpp
Comment (fjeremic) [reply compiler/optimizer/OptimizationData.hpp]: No better time than now! Fixed in [force-push](https://github.com/eclipse/omr/compare/d893d15411c7d2be105eab2b891ec734436bf948..9e907fbc46572cc952eb9db1e204f9e6c8b03d72).

File: compiler/x/codegen/SIMDTreeEvaluator.cpp
Comment (fjeremic) [reply compiler/x/codegen/SIMDTreeEvaluator.cpp:45]: I see other code which only seems to do this for 64-bit and always generates 64-bit `lea` there:
https://github.com/eclipse/omr/blob/eeb1035617c5b0beea82d81bfda9eed89e8da743/compiler/x/codegen/FPTreeEvaluator.cpp#L337-L349

Could we get someone more familiar with x86 codegen to chime in? Perhaps @0xdaryl. The code being reviewed here is used for `vload` and `vloadi` evaluators. There is no AOT related code in the above snippet for evaluating `TR::dload` for example, so why should there be one for evaluating `TR::vload`? Also why is there a `is64Bit` check for the `TR::dload` case in FPTreeEvaluator.cpp above?

File: compiler/optimizer/VPHandlers.cpp
Comment (jdmpapin) [reply compiler/optimizer/VPHandlers.cpp:4761]: `lnolz` is generated only in one place that I can see, which is to replace `invokestatic Long.numberOfLeadingZeros(J)I` [in OpenJ9's IL generator](https://github.com/eclipse-openj9/openj9/blob/0c8f6b1861f99b63b99824b0befcc5a4869c1f40/runtime/compiler/ilgen/Walker.cpp#L3718-L3719). That method returns a 32-bit integer rather than a 64-bit one, which is perfectly sensible considering that the maximum possible result is only 64. And the [IL opcode properties confirm that the result is `Int32`](https://github.com/eclipse/omr/blob/eeb1035617c5b0beea82d81bfda9eed89e8da743/compiler/il/OMROpcodes.enum#L9261-L9262) rather than `Int64` (though I have no idea why they don't specify the type of the child...)

Since the result is 32-bit, we really should create "int" constraints for it, so I think the warning needs to be fixed some other way

Comment (jdmpapin) [reply compiler/optimizer/VPHandlers.cpp:4761]: Thanks for calling attention to this one btw!

Comment (fjeremic) [reply compiler/optimizer/VPHandlers.cpp]: Fixed in b2e3cb5. Did it in a separate commit to flag it out and separate the change given the size of the PR.

**Review by dsouzai — COMMENTED** [review]
Over LGTM, minor comment. 

Also I can't edit the previous comments because of the force push. It looks like regarding the `LEARegMem` change in `compiler/x/codegen/SIMDTreeEvaluator.cpp` it doesn't really matter from an AOT pov because:
```
// Size-parameterized opcodes
//
template <OMR::InstOpCode::Mnemonic Op64, OMR::InstOpCode::Mnemonic Op32>
inline OMR::InstOpCode::Mnemonic SizeParameterizedOpCode(bool is64Bit =
#ifdef TR_TARGET_64BIT
    true
#else
    false
#endif
)
   {
   return is64Bit ? Op64 : Op32;
   }
```
we're always going to generate a 64-bit lea in a 64-bit run; the bug was fixed for 32-bit really.

File: compiler/control/CompilationController.cpp
Comment (fjeremic) [reply compiler/control/CompilationController.cpp:81]: The commit comment is correct. The issue is that `tlsFree` is a macro which can evaluate to an empty string:
https://github.com/eclipse/omr/blob/eeb1035617c5b0beea82d81bfda9eed89e8da743/compiler/infra/ThreadLocal.hpp#L118-L133

This yields code that looked like this:
```
// Before preprocessor
   if (_tlsCompObjCreated)
      tlsFree(OMR::compilation);

// After preprocessor
   if (_tlsCompObjCreated)
      ;
```

The warning is telling us the `if` statement is empty.

Comment (dsouzai) [reply compiler/control/CompilationController.cpp:81]: Ah right, it's one of those non-capitalized macros... 🙄

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
jenkins build all

**Comment by dsouzai** [comment]
AIX build failed because of infra issues.

File: cmake/modules/OmrTargetSupport.cmake
Comment (dsouzai) [reply cmake/modules/OmrTargetSupport.cmake:81]: I'm gonna go ahead with the coordinated merge. @fjeremic I think just open up the issue and fix this limitation when you can.

Comment (fjeremic) [reply cmake/modules/OmrTargetSupport.cmake:81]: Opened #6053 to track this.

**Pull request merged**
